Nodejs/Product/Nodejs/Workspace/JsFileDebugLaunchTargerProvider.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Web;
+using Microsoft.NodejsTools.Debugger.DebugEngine;
+using Microsoft.NodejsTools.Project;
+using Microsoft.NodejsTools.Telemetry;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Workspace;
+using Microsoft.VisualStudio.Workspace.Debug;
+using Microsoft.VisualStudio.Workspace.Extensions.VS.Debug;
+using Microsoft.VisualStudio.Workspace.VSIntegration.Contracts;
+
+namespace Microsoft.NodejsTools.Workspace
+{
+    [ExportVsDebugLaunchTarget(ProviderType, new[] { ".js" }, ProviderPriority.Lowest)]
+    internal sealed class JsFileDebugLaunchTargerProvider : LaunchDebugTargetProvider, IVsDebugLaunchTargetProvider, ILaunchDebugTargetProvider
+    {
+        private const string ProviderType = "{0C231F54-C3EF-4719-9CA5-102B1B63A8DC}";
+
+        [ImportingConstructor]
+        public JsFileDebugLaunchTargerProvider(SVsServiceProvider serviceProvider, IVsFolderWorkspaceService workspaceService)
+            : base(serviceProvider, workspaceService)
+        {
+        }
+
+        void IVsDebugLaunchTargetProvider.SetupDebugTargetInfo(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext)
+        {
+            var nodeExe = CheckNodeInstalledAndWarn(debugLaunchContext);
+
+            var nodeVersion = Nodejs.GetNodeVersion(nodeExe);
+            if (nodeVersion >= new Version(8, 0) || NodejsProjectLauncher.CheckDebugProtocolOption())
+            {
+                this.SetupDebugTargetInfoForInspectProtocol(ref vsDebugTargetInfo, debugLaunchContext, nodeExe);
+                TelemetryHelper.LogDebuggingStarted("ChromeV2", nodeVersion.ToString(), isProject: false);
+            }
+            else
+            {
+                this.SetupDebugTargetInfoForNodeProtocol(ref vsDebugTargetInfo, debugLaunchContext, nodeExe);
+                TelemetryHelper.LogDebuggingStarted("Node6", nodeVersion.ToString(), isProject: false);
+            }
+        }
+
+        private void SetupDebugTargetInfoForNodeProtocol(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext, string nodeExe)
+        {
+            var target = $"\"{vsDebugTargetInfo.bstrExe}\"";
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(ScriptArgsKey, out var scriptArgs) && !string.IsNullOrEmpty(scriptArgs))
+            {
+                target += " " + scriptArgs;
+            }
+            vsDebugTargetInfo.bstrArg = target;
+
+            vsDebugTargetInfo.bstrExe = nodeExe;
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(NodeArgsKey, out var nodeArgs) && !string.IsNullOrWhiteSpace(nodeArgs))
```

Comment:
>IsNullOrWhiteSpace [](start = 117, length = 18)

How do you decide whether to use `IsNullOrEmpty` or `IsNullOrWhitespace`?

Reply:
unified